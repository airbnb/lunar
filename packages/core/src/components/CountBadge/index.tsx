import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Accessibility label. */
  accessibilityLabel: string;
  /** Value to show in badge. */
  value: number;
  /** Play a waggle animation. */
  waggle?: boolean;
};

/** A badge indicator with a count. */
export class CountBadge extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    waggle: false,
  };

  badgeRef = React.createRef<HTMLDivElement>();

  componentDidUpdate(prevProps: Props) {
    const { value } = this.props;

    if (value && prevProps.value && value !== prevProps.value) {
      this.runPopAnimation();
    }
  }

  runPopAnimation() {
    if (this.badgeRef.current) {
      this.badgeRef.current.animate(
        // @ts-ignore Bug: https://github.com/Microsoft/TypeScript/issues/26073
        [
          { transform: 'scale(1)' },
          { transform: 'scale(1.1)', offset: 0.3 },
          { transform: 'scale(.95)', offset: 0.8 },
          { transform: 'scale(1)' },
        ],
        300,
      );
    }
  }

  render() {
    const { cx, accessibilityLabel, styles, value, waggle } = this.props;

    if (!value) {
      return null;
    }

    return (
      <div
        className={cx(styles.badge, waggle ? styles.animateInAndWaggle : styles.animateIn)}
        ref={this.badgeRef}
        aria-label={accessibilityLabel}
      >
        {value.toLocaleString()}
      </div>
    );
  }
}

export default withStyles(({ color, font, unit }) => {
  const waggleDuration = 300;
  const totalDuration = 2000 + waggleDuration;
  const wagglePercent = (waggleDuration / totalDuration) * 100;
  const delayPercent = 100 - wagglePercent;
  const waggleFrames = {
    name: 'waggle',
    [`${delayPercent}%`]: { transform: 'rotate(0deg)' },
    [`${wagglePercent * 0.3 + delayPercent}%`]: { transform: 'rotate(25deg)' },
    [`${wagglePercent * 0.8 + delayPercent}%`]: { transform: 'rotate(-10deg)' },
  };
  const inFrames = {
    name: 'animateIn',
    '0%': {
      transform: 'scale(0)',
      opacity: 0,
    },
    '60%': {
      transform: 'scale(1.1)',
      opacity: 1,
    },
    '80%': {
      transform: 'scale(.95)',
      opacity: 1,
    },
    '100%': {
      transform: 'scale(1)',
      opacity: 1,
    },
  };

  return {
    badge: {
      ...font.textSmall,
      backgroundColor: color.core.danger[6],
      color: color.base,
      textAlign: 'center',
      fontWeight: 'bold',
      padding: unit / 4,
      borderRadius: unit * 2,
      minWidth: unit * 4,
      boxSizing: 'border-box',
      display: 'inline-block',
    },

    animateInAndWaggle: {
      animationDuration: `500ms, ${totalDuration}ms`,
      animationName: [inFrames, waggleFrames],
      animationIterationCount: '1, infinite',
      animationDelay: '0s, 500ms',
    },

    animateIn: {
      animationDuration: '500ms',
      animationName: inFrames,
    },
  };
})(CountBadge);
