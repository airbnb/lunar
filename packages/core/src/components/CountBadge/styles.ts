import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, font, unit }) => {
  const waggleDuration = 300;
  const totalDuration = 2000 + waggleDuration;
  const wagglePercent = (waggleDuration / totalDuration) * 100;
  const delayPercent = 100 - wagglePercent;
  const waggleFrames = {
    name: 'waggle',
    [`${delayPercent}%`]: { transform: 'rotate(0deg)' },
    [`${wagglePercent * 0.3 + delayPercent}%`]: { transform: 'rotate(25deg)' },
    [`${wagglePercent * 0.8 + delayPercent}%`]: { transform: 'rotate(-10deg)' },
  };
  const inFrames = {
    name: 'animateIn',
    '0%': {
      transform: 'scale(0)',
      opacity: 0,
    },
    '60%': {
      transform: 'scale(1.1)',
      opacity: 1,
    },
    '80%': {
      transform: 'scale(.95)',
      opacity: 1,
    },
    '100%': {
      transform: 'scale(1)',
      opacity: 1,
    },
  };

  return {
    badge: {
      ...font.textSmall,
      backgroundColor: color.core.danger[6],
      color: color.base,
      textAlign: 'center',
      fontWeight: 'bold',
      padding: unit / 4,
      borderRadius: unit * 2,
      minWidth: unit * 4,
      boxSizing: 'border-box',
      display: 'inline-block',
    },

    animateInAndWaggle: {
      animationDuration: `500ms, ${totalDuration}ms`,
      animationName: [inFrames, waggleFrames],
      animationIterationCount: '1, infinite',
      animationDelay: '0s, 500ms',
    },

    animateIn: {
      animationDuration: '500ms',
      animationName: inFrames,
    },
  };
};

export default styleSheet;

export { styleSheet };
