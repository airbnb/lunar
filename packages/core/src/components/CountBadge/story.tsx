import React from 'react';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';
import CountBadge from '.';

class CountExample extends React.Component<{}, { count: number }> {
  state = { count: 5 };

  render() {
    return (
      <>
        <CountBadge
          accessibilityLabel={`${this.state.count} unread messages`}
          value={this.state.count}
        />
        <br />
        <br />
        <ButtonGroup>
          <Button
            small
            onClick={() => this.setState(prevState => ({ count: prevState.count + 1 }))}
          >
            Increment badge count
          </Button>
          <Button small onClick={() => this.setState({ count: 0 })}>
            Reset badge count
          </Button>
        </ButtonGroup>
      </>
    );
  }
}

export default {
  title: 'Core/CountBadge',
  parameters: {
    inspectComponents: [CountBadge],
  },
};

export function defaultAnimation() {
  return <CountBadge accessibilityLabel="5 unread messages" value={5} />;
}

defaultAnimation.story = {
  name: 'Default animation.',
};

export function addWaggleAnimation() {
  return <CountBadge waggle accessibilityLabel="5 unread messages" value={5} />;
}

addWaggleAnimation.story = {
  name: 'Add waggle animation.',
};

export function changeValueToTriggerAnimations() {
  return <CountExample />;
}

changeValueToTriggerAnimations.story = {
  name: 'Change value to trigger animations.',
};
