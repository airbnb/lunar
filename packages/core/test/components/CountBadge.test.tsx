import React from 'react';
import { mount } from 'enzyme';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import CountBadge from '../../src/components/CountBadge';

describe('<CountBadge />', () => {
  const props = {
    accessibilityLabel: '5 unread messages',
    value: 5,
  };

  it('renders default', () => {
    const wrapper = shallowWithStyles(<CountBadge {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('does not render when value is 0', () => {
    const wrapper = shallowWithStyles(<CountBadge {...props} value={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('handles waggle class based on interval setting', () => {
    const wrapper = shallowWithStyles(<CountBadge {...props} />);
    const oldClass = wrapper.prop('className');

    wrapper.setProps({
      waggle: true,
    });

    expect(wrapper.prop('className')).not.toBe(oldClass);
  });

  it('adds animation when value is changed', () => {
    const wrapper = mount(<CountBadge {...props} waggle />);
    const div = wrapper.find('div').getDOMNode() as HTMLDivElement;

    div.animate = jest.fn();

    wrapper.setProps({ value: 10 });

    expect(div.animate).toHaveBeenCalledTimes(1);
    expect(div.animate).toHaveBeenCalledWith(
      [
        { transform: 'scale(1)' },
        { transform: 'scale(1.1)', offset: 0.3 },
        { transform: 'scale(.95)', offset: 0.8 },
        { transform: 'scale(1)' },
      ],
      300,
    );

    wrapper.setProps({ value: 0 });
    expect(div.animate).toHaveBeenCalledTimes(1);
  });
});
