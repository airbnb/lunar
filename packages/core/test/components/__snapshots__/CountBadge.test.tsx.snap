// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountBadge /> does not render when value is 0 1`] = `""`;

exports[`<CountBadge /> renders default 1`] = `
<div
  aria-label="5 unread messages"
  className="badge_134snep-o_O-animateIn_q9sqz6"
>
  5
</div>
`;
